### ISSUE TITLE
Race Condition in reduce_first_kernel

### ISSUE THREAD

**Issue by nathanjackson** [arrayfire/arrayfire#625]
I believe I've found a race condition in the reduce_first_kernel for the CUDA backend.  On a GTX 750 Ti the following code produces a race condition that is detected by cuda-memcheck:

    #include <arrayfire.h>

    int main()
    {
        af::array data = af::randu( 3000, 1000, f32 );
        float result = af::mean<float>(data);
        return 0;
    }

I'm not sure if its a problem in the mean function or reduce_first_kernel.

Here is the result from cuda-memcheck

```
$ cuda-memcheck --tool racecheck ./test
========= CUDA-MEMCHECK
========= Race reported between Write access at 0x000009f0 in void cuda::kernel::reduce_first_kernel<float, float, af_op_t=0, unsigned int=256>(cuda::Param<float>, cuda::CParam<float>, unsigned int, unsigned int, unsigned int)
=========     and Read access at 0x00000a30 in void cuda::kernel::reduce_first_kernel<float, float, af_op_t=0, unsigned int=256>(cuda::Param<float>, cuda::CParam<float>, unsigned int, unsigned int, unsigned int) [2936 hazards]
=========
========= Race reported between Write access at 0x00000998 in void cuda::kernel::reduce_first_kernel<float, float, af_op_t=0, unsigned int=256>(cuda::Param<float>, cuda::CParam<float>, unsigned int, unsigned int, unsigned int)
=========     and Read access at 0x000009d8 in void cuda::kernel::reduce_first_kernel<float, float, af_op_t=0, unsigned int=256>(cuda::Param<float>, cuda::CParam<float>, unsigned int, unsigned int, unsigned int) [5872 hazards]
=========
========= Race reported between Write access at 0x00000948 in void cuda::kernel::reduce_first_kernel<float, float, af_op_t=0, unsigned int=256>(cuda::Param<float>, cuda::CParam<float>, unsigned int, unsigned int, unsigned int)
=========     and Read access at 0x00000988 in void cuda::kernel::reduce_first_kernel<float, float, af_op_t=0, unsigned int=256>(cuda::Param<float>, cuda::CParam<float>, unsigned int, unsigned int, unsigned int) [11744 hazards]
=========
========= Race reported between Write access at 0x00000a48 in void cuda::kernel::reduce_first_kernel<float, float, af_op_t=0, unsigned int=256>(cuda::Param<float>, cuda::CParam<float>, unsigned int, unsigned int, unsigned int)
=========     and Read access at 0x00000a88 in void cuda::kernel::reduce_first_kernel<float, float, af_op_t=0, unsigned int=256>(cuda::Param<float>, cuda::CParam<float>, unsigned int, unsigned int, unsigned int) [1468 hazards]
=========
========= RACECHECK SUMMARY: 4 hazards displayed (4 errors, 0 warnings) 
```

I don't know if this is an issue with the mean function or if these race conditions are harmless as it does seem to produce the correct results, but I'm thinking it might be possible that the results are skewed for other functions that use the reduce_first_kernel.

**Comment by pavanky** [comment]
Thanks for reporting this. I am able to reproduce this and am in the process of debugging it further.

**Comment by nathanjackson** [comment]
If it helps this also happens on my GTX 760.

**Comment by pentschev** [comment]
Fixed in #634
